File: Ruby/binary-search.rb
# Define the binary search method, that will return the index value
# of a given piece of data in a sorted array

def binary_search(array, value)
  return nil if array.length == 1 && array[0] != value
  midpoint = array.length/2
  return midpoint if array[midpoint] == value
  if array[midpoint] < value
    midpoint + binary_search(array[midpoint..array.length], value) if binary_search(array[midpoint..array.length], value)
  else
    binary_search(array[0...midpoint], value)
  end
end

# Generate a sorted array

array = [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]

# Call binary search method on sorted array

print binary_search(array, 2)
print binary_search(array, 4)
print binary_search(array, 6)
print binary_search(array, 8)
print binary_search(array, 10)
print binary_search(array, 12)
print binary_search(array, 14)
print binary_search(array, 16)
print binary_search(array, 18)
print binary_search(array, 20)
print binary_search(array, 21)
print binary_search(array, 11)

File: Javascript/binary-search.js
// Define the binary search function

binarySearch = function(array, value) {
  if (array.length === 1 && array[0] != value) {
    return null;
  }
  var midpoint = Math.floor(array.length/2);
  if (array[midpoint] === value) {
    return midpoint;
  }
  if (array[midpoint] < value) {
    if (binarySearch(array.slice(midpoint, array.length), value)) {
      return midpoint + binarySearch(array.slice(midpoint, array.length), value);
    }
  } else if (array[midpoint] > value) {
    return binarySearch(array.slice(0, midpoint), value);
  }
}

// Generate a sorted array

var array = [2, 4, 6, 8, 10, 12, 14, 16, 18];

// Call binary search function on sorted array

console.log(binarySearch(array, 2));
console.log(binarySearch(array, 4));
console.log(binarySearch(array, 6));
console.log(binarySearch(array, 8));
console.log(binarySearch(array, 10));
console.log(binarySearch(array, 12));
console.log(binarySearch(array, 14));
console.log(binarySearch(array, 16));
console.log(binarySearch(array, 18));